### DISCUSSION TITLE
How to configure Fetchmail with a remote IMAP server?

### DISCUSSION THREAD

**Discussion by Doni49** [docker-mailserver/docker-mailserver#3994] [Q&A] [answered]
For the past couple decades, I have had a hosted cPanel web account that includes an email server.  This has been working well -- but I have been dealing with storage capacity issues.  My goal is to have docker-mailserve & fetchmail connect to the remote server, download messages and present them to clients that connect locally to dms.  DMS will never be exposed to the public. 

I have docker-mailserve installed and fetchmail is enabled.  I have a user created in docker-mailserve and have successfully sent and received a test message to and from this account.

But I don't see how to specify how to connect to the remote server.  

I've looked through the fetchmail documentation.  

1. But I'm not sure where to put any of the fetchmail settings I need to use.  
2. I'm also unsure how to tie the remote user messages to the local user inbox.  i.e. download messages from RemoteUserInbox and place them in LocalUserInbox?  As of right now, I'm the only user.  But at some point in the future, I may need to add additional users so knowing this will be even more of an issue.
3. The remote server is IMAP compatible.  I like using IMAP because it results in near instant arrival of messages.  I'd like to connect to the remote server via IMAP, download the messages and delete them from the server.  Is that feasible?

**Comment by glime-ai** [comment]
**Please note** that this response was created using generative AI, leveraging previously available information from the repository. This information may be helpful. However, it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To configure Fetchmail with Docker Mailserver (DMS) to connect to a remote server, download messages, and manage them locally, you need to follow these steps:

1. **Configure Fetchmail Settings:**
   - You need to create a `fetchmail.cf` configuration file. This file should be placed in the `docker-data/dms/config/` directory of your Docker Mailserver setup.
   - An example configuration for an IMAP server (like your remote server) would look like this:

     ```plaintext
     poll imap.remote-server.com proto imap
       user 'remoteuser@example.com'
       pass 'password'
       is 'localuser@yourdomain.com'
       ssl
       keep
     ```

   - In this configuration:
     - `poll imap.remote-server.com` specifies the IMAP server of your remote email.
     - `user 'remoteuser@example.com'` and `pass 'password'` are the credentials for the remote email account.
     - `is 'localuser@yourdomain.com'` specifies the local DMS user where the emails will be delivered.
     - `ssl` ensures that the connection to the IMAP server is encrypted.
     - `keep` is optional and if included, Fetchmail will leave a copy of the emails on the remote server. If you want to delete them after fetching, remove this line.

2. **Tie Remote User Messages to Local User Inbox:**
   - The `is 'localuser@yourdomain.com'` part of the Fetchmail configuration ensures that emails fetched from the remote server are delivered to the specified local user's inbox in DMS.

3. **Setting Up IMAP for Instant Message Arrival and Deletion:**
   - Fetchmail supports IMAP with options to delete messages after fetching if you do not include the `keep` command in your configuration. This setup will download the messages and then delete them from the remote server, as per your requirement.

4. **Adding Additional Users in the Future:**
   - To add more users, you would replicate the user-specific block in the `fetchmail.cf` file for each new remote-local user pair, adjusting the `user`, `pass`, and `is` fields accordingly.

After setting up your `fetchmail.cf`, ensure that the Fetchmail service is enabled in your Docker Mailserver by setting the environment variable `ENABLE_FETCHMAIL` to `1` in your `compose.yaml` file.

This setup should allow you to achieve your goal of connecting to a remote IMAP server, downloading messages, and managing them locally without exposing DMS to the public.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=0d5b86fb7eea4a54acac405212c29aba&utm_medium=github).

**Reply by Doni49** [reply]
Unfortunately, I did that but none of the remote messages are getting downloaded.  I thought I misread how to do it so I asked here.

**Reply by Doni49** [reply]
I've got the polling time set to 120 seconds. But even when I send a test message to the remote server (_which I can see the message is there if login via webmail -- but don't click is to avoid having it marked as read_) and wait an hour or more, it doesn't show up in the local server.

Here is my `fetchmail.cf` file (_remote domain name and remote username/password have been changed for posting here_):

```
poll 'mail.example.com' proto imap
  user 'user@example.com'
  pass 'pass'
  is 'don@home.arpa'
  ssl
  --nokeep
```

Here is my `compose.yaml` file for docker-mailserver:

```yaml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: docker.home.arpa
    env_file: mailserver.env
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/latest/config/security/understanding-the-ports/
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS, Authentication is DISABLED => use port 465/587 instead)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
    volumes:
      - /mnt/docker/mailserver/data/:/var/mail/
      - /mnt/docker/mailserver/state/:/var/mail-state/
      - /mnt/docker/mailserver/logs/:/var/log/mail/
      - /mnt/docker/mailserver/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
    restart: always
    stop_grace_period: 1m
    # Uncomment if using `ENABLE_FAIL2BAN=1`:
    # cap_add:
    #   - NET_ADMIN
    healthcheck:
      test: "ss --listening --tcp | grep -P 'LISTEN.+:smtp' || exit 1"
      timeout: 3s
      retries: 0
```

And Here's a screenshot taken from portainer showing the environment variables that are relevant to fetchmail:

![image](https://github.com/docker-mailserver/docker-mailserver/assets/14239245/18a0acf0-1d81-40e2-a52a-b7b6e5dd2960)

**Comment by polarathene** [comment, marked as answer]
## Local test between two DMS with fetchmail

Here you go, this will let you try the feature locally with two instances of DMS.

```yaml
# compose.yaml
services:
  dms-fetch:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.example.test
    environment:
      ENABLE_FETCHMAIL: 1
      # We change this setting to 10 for quicker testing:
      FETCHMAIL_POLL: 10
    # Link the DNS lookup `remote.test` to resolve to the `dms-remote` container IP (for `@remote.test` address):
    # This is only for this example, since no real DNS service is configured, this is a Docker internal DNS feature:
    links:
      - "dms-remote:remote.test"
    # NOTE: Optional, You only need to publish ports if you want to verify via your own mail client.
    #ports:
    #  - "465:465"  # ESMTP (implicit TLS)
    #  - "993:993"  # IMAP4 (implicit TLS)
    # You'd normally use volumes here, for simplicity of the example, all config is contained within `compose.yaml`:
    configs:
      - source: dms-accounts-fetch
        target: /tmp/docker-mailserver/postfix-accounts.cf
      - source: fetchmail
        target: /tmp/docker-mailserver/fetchmail.cf

  dms-remote:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.remote.test
    environment:
      # Allows for us send a test mail easily by trusting any mail client run within this container (`swaks`):
      PERMIT_DOCKER: container
      # Alternatively, trust and accept any mail received from clients in same subnet of dms-fetch:
      #PERMIT_DOCKER: connected-networks
    configs:
      - source: dms-accounts-remote
        target: /tmp/docker-mailserver/postfix-accounts.cf

# Using the Docker Compose `configs.content` feature instead of volume mounting separate files.
# NOTE: This feature requires Docker Compose v2.23.1 (Nov 2023) or newer:
# https://github.com/compose-spec/compose-spec/pull/446
configs:
  fetchmail:
    content: |
      poll 'mail.remote.test' proto imap
        user 'jane.doe@remote.test'
        pass 'secret'
        is 'john.doe@example.test'
        no sslcertck

  # DMS requires an account to complete setup, configure one for each instance:
  # NOTE: Both accounts are configured with the same password (SHA512-CRYPT hashed), `secret`.
  dms-accounts-fetch:
    content: |
      john.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.

  dms-accounts-remote:
    content: |
      jane.doe@remote.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.
```

```console
$ docker compose up

# John has no mail in his inbox (Your local DMS instance with fetchmail)
$ docker compose exec -it dms-fetch doveadm mailbox status -u john.doe@example.test messages INBOX
INBOX messages=0

# Jane has no mail in her inbox (Your remote DMS instance)
$ docker compose exec -it dms-remote doveadm mailbox status -u jane.doe@remote.test messages INBOX
INBOX messages=0

# Send the remote DMS a mail internally (PERMIT_DOCKER ENV allows to send this without authentication):
$ docker compose exec -it dms-remote swaks --server localhost --from someone@example.test --to jane.doe@remote.test

# Jane has mail now, until fetchmail takes it:
$ docker compose exec -it dms-remote doveadm mailbox status -u jane.doe@remote.test messages INBOX
INBOX messages=1

# Wait for fetchmail poll to pass, and check inboxes again,
# Fetchmail has retrieved Janes mail, deleting from her inbox, sending the copy to John's inbox:
$ docker compose exec -it dms-remote doveadm mailbox status -u jane.doe@remote.test messages INBOX
INBOX messages=0
$ docker compose exec -it dms-fetch doveadm mailbox status -u john.doe@example.test messages INBOX
INBOX messages=1
```

---

## Fetchmail config

- `--nokeep` is not what you want for the `fetchmail.cf` file, that is for the CLI daemon options, the equivalent [config keyword](https://www.fetchmail.info/fetchmail-man.html#toc-entry-37) is `no keep` (default).
- Likewise `ssl` should be the default AFAIK. As my example above was testing locally, extra work is required to setup TLS certs that can be verified, so to skip that additional complexity I used `no sslcertck`. Careful not to have this line in production config, as you will likely want to verify trust is correct.

You can see that it works as intended. So please refer to this example and adapt it to your remote mail server. It should successfully be able to retrieve the mail. Then adapt the example to your actual mail server setup, and it should help you troubleshoot where the problem is 👍 

Looking at [our docs for fetchmail](https://docker-mailserver.github.io/docker-mailserver/v13.3/config/advanced/mail-fetchmail/#configuration) config, the settings are effectively the same in the example above as they are in the docs example. I don't see anything that I can update there, beyond mentioning this more complete example for local testing/verification?

---

Docs also provide some debug advice: https://docker-mailserver.github.io/docker-mailserver/v13.3/config/advanced/mail-fetchmail/#debugging

For any additional trobleshooting, be sure to check the logs from DMS. When Fetchmail has problems connecting, it should be logging messages like this:

```
dms-fetch   | May  2 13:50:25 mail fetchmail[653]: remote.test: upgrade to TLS failed.
dms-fetch   | May  2 13:50:25 mail fetchmail[653]: Unknown login or authentication error on jane.doe@remote.test
dms-fetch   | May  2 13:50:25 mail fetchmail[653]: socket error while fetching from jane.doe@remote.test
dms-fetch   | May  2 13:50:25 mail fetchmail[653]: Query status=2 (SOCKET)
dms-remote  | May  2 13:50:25 mail dovecot: imap-login: Disconnected: Connection closed (no auth attempts in 0 secs): user=<>, rip=172.19.0.3, lip=172.19.0.2, session=<rmao024XPqKsEwAD>
```

As you can see, the connection to `mail.remote.test` was partially successful, but the TLS connection was not, which caused the actual login/auth error for `jane.doe@remote.test`. This was when I tried with some self-signed certs locally, but `SSL_TYPE=self-signed` ENV or `SSL_TYPE=manual` wasn't happy verifying certs as the root CA used wasn't trusted within the container 😓

**Reply by Doni49** [reply]
Thanks for the detailed response!  It is greatly appreciated.  I have a few follow-up questions about your answer.
1. I'm not sure what you mean by "separate files" here -- separate files for each message?:
`# Using the Docker Compose `configs.content` feature instead of volume mounting separate files.`
2. It looks like you have this in the compose.yaml file.  From my reading of the docs, I thought it needed to be in fetchmail.cf:
```
fetchmail:
    content: |
      poll 'mail.remote.test' proto imap
        user 'jane.doe@remote.test'
        pass 'secret'
        is 'john.doe@example.test'
        no sslcertck
```
3. Based on this portion of your example, it sounds like put this in a cron entry.  Again based on my reading of the docs, I thought the polling time was the interval and it the poll would happen automatically based on the the cofiguration in fecthmail.cfg:
`# Jane has no mail in her inbox (Your remote DMS instance)
$ docker compose exec -it dms-remote doveadm mailbox status -u jane.doe@remote.test messages INBOX
INBOX messages=0`

**Reply by polarathene** [reply]
> I'm not sure what you mean by "separate files" here

**TL;DR:** Exactly as the comment implies:
- Anything indented below a `content: |` YAML line is the contents of a file that you'd normally have outside of `compose.yaml`, instead of a filename these have an identifier in the YAML config like `configs.fetchmail`.
- A service like `dms-fetch` then refers to that content with `configs.source: fetchmail` which is mounted into the container at the `configs.target: /tmp/dockermailserver/fetchmail.cf` path, creating a file inside the container with that source content.

If that's still confusing, here's a more verbose explanation:

The `compose.yaml` I provided is a full demo you can run. Normally you would provide your own separate config files that are provided to the container via the `volumes` property for a service/container defined in `compose.yaml`, but in this example instead of `volumes` I'm using `configs`.

A separate section in `compose.yaml` called `configs` defines several files with content provided inline. This is just an easy way for small examples like this to be self-contained into a single `compose.yaml` file that you can copy/paste to run, instead of making the individual files. I would generally advise separate files for production use, but you can use either approach.

Thus if using `volumes`, you would have a `fetchmail.cf` file as per our docs, and the contents would be the equivalent without the extra indentation needed in YAML:

```
# Volume mounted into the container as `/tmp/docker-mailserver/fetchmail.cf`:
poll 'mail.remote.test' proto imap
  user 'jane.doe@remote.test'
  pass 'secret'
  is 'john.doe@example.test'
  no sslcertck
```

In my `compose.yaml` example, you can see that the `fetchmail` content config is referenced by the `dms-fetch` service via it's own `configs` property. I reference the config name `fetchmail` as the source, and set the path as the `target`. Hopefully that makes more sense with how it works.

The `configs` approach is static. So it would not really be suitable for files like `postfix-accounts.cf` where you may update it and make changes with our `setup` command. Which is why `volumes` is better to mount a local folder of config files to `/tmp/docker-mailserver/`.

---

> Based on this portion of your example, it sounds like put this in a cron entry.

No, cron is for scheduling jobs, no need for that. Just configure the `fetchmail.cf` file as described. I rewrote our fetchmail docs after answering this, it will be part of the v14 release, but for now you may [view the update on our edge published docs](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/mail-fetchmail/) (_direct link to fetchmail page_). That may be more clearer for you to follow?

> I thought the polling time was the interval and it the poll would happen automatically based on the the configuration in `fecthmail.cf`

Yes, you can set the polling interval via an ENV we provide. I did this in the `compose.yaml` example above, setting it to 10 seconds for quick feedback when testing. Try the commands as I provided, when an email is sent to the remote mail server for a user, fetchmail polls at that interval and if new mail is found, it will retrieve it and remove from the remote inbox, while delivering it to the local DMS user configured.

The commands I provided (_the commands are prefixed with `$` to indicate this line is a command you'd type after the `$`_), show the expected output after the command. In this case it looks like `INBOX messages=1` where the `1` tells me that there is now 1 message in that users inbox, when previously it was `0`. If the polling is too frequent when trying this example, just adjust it to a longer interval, and you can better observe this change between the local and remote accounts.

**Reply by Doni49** [reply]
No questions from me -- just trying to answer what seemed like questions from you...

> > I'm not sure what you mean by "separate files" here

> Exactly as the comment implies:"

I've heard of mail servers that use individual folders to represent a folder within the user's folder on the network and individual files within the folders.  On the other hand, I've heard of one large file for each folder.

> No, cron is for scheduling jobs, no need for that. Just configure the fetchmail.cf file as described.

The code in your example (docker exec -it....) looked like code I would be running at the command line.  The only way I could think to automate that was cron.  

I'll take a deep dive into your example  -- probably work on it tomorrow.

**Reply by polarathene** [reply]
> I've heard of mail servers that use individual folders to represent a folder within the user's folder on the network and individual files within the folders. On the other hand, I've heard of one large file for each folder.

This is unrelated to what I was discussing / demonstrating. Mail storage is [handled in `/var/mail` as we document](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/optional-config/) (_linking to edge, as the upcoming v14 release docs is much better on this topic_) and [show in our example `compose.yaml`](https://github.com/docker-mailserver/docker-mailserver/blob/e2c2a22dcf2d8adc12f303b84b694945e8dc072e/compose.yaml#L16-L17):

https://github.com/docker-mailserver/docker-mailserver/blob/e2c2a22dcf2d8adc12f303b84b694945e8dc072e/compose.yaml#L16-L17

You'll find that in `/var/mail` there is a folder per domain, and that has sub folders per user. So `john.doe@example.com` is at `/var/mail/example.com/john.doe/`, this storage is managed by the Dovecot service running in the container.

You are right, there is various formats that Dovecot supports (_we do [support changing this](https://github.com/docker-mailserver/docker-mailserver/blob/e2c2a22dcf2d8adc12f303b84b694945e8dc072e/test/tests/parallel/set1/dovecot/mailbox_format_dbox.bats#L4-L9)_), but for the most part you shouldn't need to worry about it. Just provide the volume mount as shown in the linked `compose.yaml` example. That will persist the mail storage to your host disk. Your mail client like Thunderbird will then connect to Dovecot, and Dovecot will provide the mail access via IMAP or POP3 protocols to the mail client, how Dovecot manages the mail data is rarely important for you the user :)

---

> The code in your example (docker exec -it....) looked like code I would be running at the command line. The only way I could think to automate that was cron.

I think you misunderstand what I've shown you.

You take the example `compose.yaml` config, then in a terminal at that files location, run the commands to follow along with the example and expected output from commands. As you can see it starts with `docker compose up`, then `docker compose exec ...` commands follow to run commands in the two containers to count the total mail in the queried users inbox.

---

> I'll take a deep dive into your example -- probably work on it tomorrow.

Sure! It's fairly quick to run, just copy/paste and save as `compose.yaml`, then in a terminal run the commands like I mentioned. That's all there is to it 👍 

Once you've done that, then you have a reference example to try for troubleshooting whatever is not working on your end.
- A pretend "remote" container for testing (_you can use the `swaks` command I showed as an easy and quick way to add a mail to this "remote" container_).
- Your own DMS config can then use the `fetchmail.cf` from the example, and it should work correctly just like the example `dms-fetchmail` container does with `dms-remote` container.
- Once that's confirmed as working, then adjust the `fetchmail.cf` to fetch from your real remote mail server instead of `dms-remote` container and again, it should work, otherwise you'll have identified where the problem is to investigate further. Don't forget to remove the `no sslcertck` line (_only exists to make the example simpler by skipping the need for TLS certs locally_).

**Reply by Doni49** [reply]
> Sure! It's fairly quick to run, just copy/paste and save as compose.yaml, then in a terminal run the commands like I mentioned.

When you first posted, I didn't realize your code was linked to a live server -- I thought it was just example code that I'd use to modify to match my setup.  So I did as you suggested thinking that once it was up and running, I'd modify as needed.

I pasted your example into a new compose.yaml file.  As soon as I ran docker-compose up, I was greeted with three error messages: 
```
ERROR: The Compose file './compose.yaml' is invalid because:
configs.dms-accounts-fetch value 'content' does not match any of the regexes: '^x-'
configs.dms-accounts-remote value 'content' does not match any of the regexes: '^x-'
configs.fetchmail value 'content' does not match any of the regexes: '^x-'
doni49@HomeRasPi5:/mnt/docker/dms $

```
I don't see any regex entries to be able to check them.

**Reply by polarathene** [reply]
> When you first posted, I didn't realize your code was linked to a live server

The `compose.yaml` will run two DMS containers locally, it should work completely offline 👍 

> As soon as I ran docker-compose up, I was greeted with three error messages

Please note the comment above the `configs:` top-level section. If your Docker Compose install does not recognize the `content` key, it is because you do not have the minimum version of Docker Compose required to understand it. You need at least v2.23.1 (released Nov 2023).

If you visit the official Docker docs, they have guidance for installing the latest version on your OS. If you do not want to upgrade your Docker Compose, then you will need to adjust the `compose.yaml` to use `volumes:` instead of `configs:`. This will also require you to create three files:

**`compose.yaml`:**
```yaml
# Update the services below to use `volumes` instead of `configs`:

# dms-fetch
volumes:
  - ./dms-fetch/fetchmail.cf:/tmp/docker-mailserver/fetchmail.cf
  - ./dms-fetch/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf

# dms-remote
volumes:
  - ./dms-remote/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf
```

**`dms-fetch/fetchmail.cf`:**
```
poll 'mail.remote.test' proto imap
  user 'jane.doe@remote.test'
  pass 'secret'
  is 'john.doe@example.test'
  no sslcertck
```

**`dms-fetch/postfix-accounts.cf`:**
```
john.doe@example.test|{SHA512-CRYPT}$6$sbgFRCmQ.KWS5ryb$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.
```

**`dms-remote/postfix-accounts.cf`:**
```
jane.doe@remote.test|{SHA512-CRYPT}$6$sbgFRCmQ.KWS5ryb$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.
```

The `postfix-account` values have been slightly adjusted. When in `compose.yaml` the special character `$` needed to be repeated to escape it. When used as a volume mount file, no escaping is needed, so all `$$` were changed to `$`.

Now with that modification, the example should work on older Docker Compose 👍

**Reply by Doni49** [reply]
I didn't get any error messages about the #compose.yaml line.  But as far as docker version goes, I installed it & docker-compose last week and then ran apt update and apt upgrade both.  I even tried update and upgrade again before posting my reply just to be sure.

I just checked the versions of both docker and docker-compose.
```
don@DockerServer:~ $ docker-compose -v
docker-compose version 1.29.2, build unknown
don@DockerServer:~ $ docker -v
Docker version 20.10.24+dfsg1, build 297e128
don@DockerServer:~ $
```
I'll try it with your revised code.

**Reply by Doni49** [reply]
I used the compose.yaml from your first example, removed the fetchmail configs portion & put in the new config.yaml code from your last message.

```
# compose.yaml
services:
  dms-fetch:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.example.test
    environment:
      ENABLE_FETCHMAIL: 1
      # We change this setting to 10 for quicker testing:
      FETCHMAIL_POLL: 10
    # Link the DNS lookup `remote.test` to resolve to the `dms-remote` container IP (for `@remote.test` address):
    # This is only for this example, since no real DNS service is configured, this is a Docker internal DNS feature:
    links:
      - "dms-remote:remote.test"
    # NOTE: Optional, You only need to publish ports if you want to verify via your own mail client.
    #ports:
    #  - "465:465"  # ESMTP (implicit TLS)
    #  - "993:993"  # IMAP4 (implicit TLS)
    # You'd normally use volumes here, for simplicity of the example, all config is contained within `compose.yaml`:
    configs:
      - source: dms-accounts-fetch
        target: /tmp/docker-mailserver/postfix-accounts.cf
      - source: fetchmail
        target: /tmp/docker-mailserver/fetchmail.cf

  dms-remote:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.remote.test
    environment:
      # Allows for us send a test mail easily by trusting any mail client run within this container (`swaks`):
      PERMIT_DOCKER: container
      # Alternatively, trust and accept any mail received from clients in same subnet of dms-fetch:
      #PERMIT_DOCKER: connected-networks
    configs:
      - source: dms-accounts-remote
        target: /tmp/docker-mailserver/postfix-accounts.cf
# Update the services below to use `volumes` instead of `configs`:

# dms-fetch
volumes:
  - ./dms-fetch/fetchmail.cf:/tmp/docker-mailserver/fetchmail.cf
  - ./dms-fetch/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf

# dms-remote
volumes:
  - ./dms-remote/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf
```

Now I'm getting a new error:
```
don@DockerServer:/mnt/docker/dms $ docker-compose up
ERROR: In file './compose.yaml', volumes must be a mapping, not an array.
don@DockerServer:/mnt/docker/dms $

```

**Reply by polarathene** [reply]
@Doni49 I think you are quite new to Docker in general?

Sorry if it wasn't clear, but as the comment I added to the `compose.yaml` config adjustment states, you need to add the `volumes:` parts to the `dms-fetch` and `dms-remote` services from the original, not directly after. Make sure you indent them correctly. They are meant to replace the `configs:` section of those services, like instructed... actually here you go:

```yaml
services:
  dms-fetch:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.example.test
    environment:
      ENABLE_FETCHMAIL: 1
      FETCHMAIL_POLL: 10
    links:
      - "dms-remote:remote.test"
    volumes:
      - ./dms-fetch/fetchmail.cf:/tmp/docker-mailserver/fetchmail.cf
      - ./dms-fetch/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf

  dms-remote:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    hostname: mail.remote.test
    environment:
      PERMIT_DOCKER: container
    volumes:
      - ./dms-remote/postfix-accounts.cf:/tmp/docker-mailserver/postfix-accounts.cf
```

I've removed the commentary from this `compose.yaml` so you can just focus on what we're doing. If you need context, please refer to the earlier `compose.yaml` content shared.

---

> I just checked the versions of both docker and docker-compose.
> 
> ```
> don@DockerServer:~ $ docker-compose -v
> docker-compose version 1.29.2, build unknown
> don@DockerServer:~ $ docker -v
> Docker version 20.10.24+dfsg1, build 297e128
> don@DockerServer:~ $
> ```

These are very old... They're from 2020. You probably want to refer to [Dockers install docs](https://docs.docker.com/engine/install/) to get a newer version. If you're on Ubuntu [here is the guidance](https://docs.docker.com/engine/install/ubuntu/) for that distro.

The above `compose.yaml` I provided in this reply should still work with your current Docker Compose, but if you have other technical troubles I would encourage you to get something more recent. DMS does not officially support Docker from 4 years ago.

**Reply by Doni49** [reply]
> I think you are quite new to Docker in general?

That would be an understatement.  I've been watching YT videos on the topic for about a month or so knowing that I wanted to use it.  But last weekend was my first time actually trying it out.

> Sorry if it wasn't clear, but as the comment I added to the compose.yaml config adjustment states, you need to add the volumes: parts to the dms-fetch and dms-remote services from the original, not directly after. Make sure you indent them correctly. They are meant to replace the configs: section of those services, like instructed... actually here you go:

I think I get what you were trying to explain now.  Your compose.yaml file just creates two separate brand new containters.  One is the "local mailserver" like I was trying to get setup and the other is intended to simulate the remote connection to which I want to connect (my hosted cPanel server).

I ran your code and it didn't present any errors.

**Reply by polarathene** [reply]
> I think I get what you were trying to explain now.

Yes that's right! 😁 

Two containers are configured, both running DMS. One is meant to represent your local DMS container with fetchmail, the other the other mail server you want fetchmail to collect from.

> I ran your code and it didn't present any errors.

Fantastic! 😎 🎉 

Now you can continue with the earlier guidance.
1. Adjust the example `fetchmail.cf` to connect to your real remote mail server (you may want to add the line `keep` for testing if you don't want it to remove the mail from your real remote just yet. **Alternatively, skip to the next step** if it's easier for you to inspect mail.
2. Then try that `fetchmail.cf` on your real local DMS (either with the fake or real remote server)
   - If it's with the fake, you'll need the `dms-remote` service added to your real `compose.yaml` and the `link` from `dms-fetch` setting added to your real DMS container.
   - If you have the `fetchmail.cf` adjusted for your real remote mail server, then hopefully everything is working now 👍 (**this may be the easiest for you to try first?**) Don't forget to remove the `no sslcertck` line from the config once you're successful.
3. If you're not having success, then you'll need to experiment a bit with the information / guidance I've provided to troubleshoot what is misconfigured.

**Reply by Doni49** [reply]
I modified fetchmail.cf to download from my real remote mail server.  The test message that I sent is no longer there.  So it appears that it should be waiting for me on the local server.  I don't have an email client configured to check the test account that was created locally.  But since I was previously able to connect both my android phone and thunderbird on my laptop to the local server, I'm assuming that should be a simple process.

Thanks for all your help!  It has been greatly appreciated!

**Reply by polarathene** [reply]
> So it appears that it should be waiting for me on the local server.

Yes, although in my `compose.yaml` example there is no `/var/mail` volume mount like you'd usually have. This means as soon as the container is removed so is that data. Hence why I mentioned adding `keep` to `fetchmail.cf` when testing, which would prevent deleting the original mail from the remote server, once you've got it working as you want you can remove that line and the default behaviour to delete the original mail when fetched will occur going forward.

---

> I don't have an email client configured to check the test account that was created locally

Thunderbird can do this, but it's probably not something I have time to walk you through (_especially with my offline `compose.yaml` example with two mailservers_) for local container connections.

This is why my example commands showed you how to check a mail had arrived:

```console
# This is checking `john.doe@example.test` account as defined in `postfix-accounts.cf`, for the example dms-fetch container:
$ docker compose exec -it dms-fetch doveadm mailbox status -u john.doe@example.test messages INBOX
# The output from the command should look something like this:
INBOX messages=1
```

That's an easy way to confirm the email is received locally at least, since you can see the count increase each time.

There's other commands to access the actual email contents, but I don't recall what those are. The mail is stored with filenames that aren't very human friendly to identify/type 😓 

If you have `/var/mail` mounted, you can possibly browse and inspect file contents to find the raw mail data, which a mail client will parse to display.

---

> But since I was previously able to connect both my android phone and thunderbird on my laptop to the local server, I'm assuming that should be a simple process.

Oh, if you've got separate client devices that's probably easier yes 👍 My `compose.yaml` example omitted publishing any ports, which you'd normally do for them to be available from the Docker host or any devices that can connect to that server. Then you'd be able to connect with Thunderbird or any other mail client.

---

> Thanks for all your help! It has been greatly appreciated!

You're welcome!

Thanks for giving DMS a go and helping to improve our documentation on fetchmail 😎 (_the next release of DMS has our fetchmail docs page link to this discussion to help others troubleshoot config_)

**Reply by Doni49** [reply]
I added my own user account to the local server.  Then I changed the fetchmail.cf configuration so that mail in the remote inbox goes to the new account.  I've got thunderbird and my phone configured to get messages from the local server.  The test messages are arriving!  <happy dance>  :)

Oh -- and I added the necessary ports to allow email clients to connect to the local server.

**Reply by polarathene** [reply]
**EDIT:** Seems like you deleted what I was replying to before I sent this, I take it you figured it out 😎 

> Can I safely have fetchmail.cf check TWO remote servers and download messages from BOTH inboxes into ONE local account?

AFAIK that should be possible. View the current `edge` release docs (these will be v14 when released): https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/mail-fetchmail/

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/1e356a19-9766-41e3-8420-9f24cfff97d1)

So something like this should work as your `fetchmail.cf` based on what we've used in the above discussion:

```
poll 'mail.remote.test' proto imap
  user 'jane.doe@remote.test' pass 'secret' is 'john.doe@example.test'

poll 'mail.another-remote.test' proto imap
  user 'alice@another-remote.test' pass 'secret' is 'john.doe@example.test'
```
